src/ApiController.php

```diff
@@ -0,0 +1,284 @@
+            $emoji = $this->db->emoji();
+
+            //check if token is valid then proceed or return an error message
+            if (Token::isValid($token, $this->db)) {
+                //get post data and make some additions
+                $all_post_vars = $this->app->request->post();
+                $all_post_vars['date_created'] = $this->get_current_time('Y-m-d H:i:s');
+                $all_post_vars['date_modified'] = $this->get_current_time('Y-m-d H:i:s');
+                $all_post_vars['created_by'] = Token::owner($token, $this->db);
+
+                //insert record
+                $res = $emoji->insert($all_post_vars);
+                $timestamp = $res['date_created'];
+
+                //return result of this request operation and the inserted emoji
+                echo json_encode([
+                    'message' => 'Emoji successfully added!',
+                    'emoji'   => $emoji->where('date_created = ?', $timestamp)->fetch(),
+                    ]);
+            } else {
+                $this->app->response->setStatus(401);
+                echo json_encode(['message' => 'Invalid Token!']);
+            }
+        };
+    }
+
+    /**
+     * Controller action that processes the request made to the update endpoint via either put or patch.
+     *
+     * @return closure outputs a status message 
+     */
+    public function updateEmoji()
+    {
+        return function ($id) {
+            $this->app->response->headers->set('Content-Type', 'application/json');
+            $token = $this->app->request->headers->get('user-token');
+
+            if (Token::isValid($token, $this->db) && $this->user_owns_emoji($token, $id) && $this->emoji_exists($id)) {
+                $all_post_vars = $this->app->request->post();
+                $all_post_vars['date_modified'] = $this->get_current_time('Y-m-d H:i:s');
+
+                $emoji = $this->db->emoji('id = ?', $id);
+
+                if ($emoji->update($all_post_vars)) {
+                    echo json_encode(['message' => 'Smiley updated successfully']);
+                } else {
+                    $this->app->response->setStatus(304);
+                    echo json_encode(['message' => 'Smiley could not be updated']);
+                }
+            } else {
+                $this->app->response->setStatus(401);
+                echo json_encode(['message' => 'At least one of the authentication requirements failed']);
+            }
+        };
+    }
+
+    /**
+     * Controller action that processes the request made to the delete endpoint.
+     *
+     * @return closure outputs a status message
+     */
+    public function deleteEmoji()
+    {
+        return function ($id) {
+            $this->app->response->headers->set('Content-Type', 'application/json');
+            $token = $this->app->request->headers->get('user-token');
+            if (Token::isValid($token, $this->db) && $this->user_owns_emoji($token, $id) && $this->emoji_exists($id)) {
```

Comment:
**Code Quality -- Nitpick**
This logic is repeated all through-out, the standard way is to have this implemented as a [middleware](http://docs.slimframework.com/routing/middleware/).